source/common/upstream/upstream_impl.cc

```diff
@@ -1575,7 +1577,15 @@ bool BaseDynamicClusterImpl::updateDynamicHostList(const HostVector& new_hosts,
   if (!current_priority_hosts.empty() && dont_remove_healthy_hosts) {
     erase_from =
         std::remove_if(current_priority_hosts.begin(), current_priority_hosts.end(),
-                       [&updated_hosts, &final_hosts, &max_host_weight](const HostSharedPtr& p) {
+                       [&all_new_hosts, &new_hosts_for_current_priority, &updated_hosts,
+                        &final_hosts, &max_host_weight](const HostSharedPtr& p) {
+                         if (all_new_hosts.contains(p->address()->asString()) &&
+                             !new_hosts_for_current_priority.contains(p->address()->asString())) {
+                           // It'll be an immediate deletion if the host exists in the cluster
+                           // somewhere other than this priority.
+                           return false;
+                         }
+
```

Comment:
I thought I'd already commented on this but seems like not: currently we somewhat support having a single endpoint exist in multiple priorities at once (I know of at least one company that told me that they did this). Will this break whatever support we have for having an endpoint exist in multiple priorities?

Reply:
That might have been me, we wanted to do that ages ago, but it doesn't work because the existing `updated_hosts` check (upstream_impl.cc line 1452) means that it only gets added to the first priority.

Reply:
I think @bplotnick was using endpoints in multiple priorities at a previous company as well, but maybe this functionality broke at some point?

Reply:
There's a lot of context including the use-case and tradeoffs of this strategy here: https://github.com/envoyproxy/envoy/issues/4280. I left Yelp a little over a year ago and we were still using this approach.

CC: @kaisen

Reply:
I think #3959 may have changed that behaviour. The address is added to `updated_hosts` when processing the first priority it's seen in. In further priorities a host with the same address is skipped over as it's already in `updated_hosts`, and so it's not added to `final_hosts` for those priorities.

Reply:
Unless I've missed some trick to doing it, I've confirmed confirmed that the ability to have an address in multiple priorities broke quite some time ago. It gets ignored from all but the first priority it appears in. Fixing that would require significant rewriting of this code.

As a result, I do not believe that this PR breaks behaviour that anyone might be relying on.

Reply:
Thanks for digging into it, I think I agree with your conclusion so I'm fine with going forward with this change. @mattklein123 any concerns on your end with this change?